sampling/include/ebpps_sample_impl.hpp

```diff
@@ -0,0 +1,535 @@
+      set_partial(conditional_forward<FwdSample>(*other.partial_item_));
+    }
+  }
+}
+
+template<typename T, typename A>
+string<A> ebpps_sample<T ,A>::to_string() const {
+  std::ostringstream oss;
+  oss << "   sample:" << std::endl;
+  uint32_t idx = 0;
+  for (const T& item : data_)
+    oss << "\t" << idx++ << ":\t" << item << std::endl;
+  oss << "   partial: " << (bool(partial_item_) ? (*partial_item_) : "NULL") << std::endl;
+
+  return oss.str();
+}
+
+template<typename T, typename A>
+void ebpps_sample<T,A>::subsample(uint32_t num_samples) {
+  // we can perform a Fisher-Yates style shuffle, stopping after
+  // num_samples points since subsequent swaps would only be
+  // between items after num_samples. This is valid since a
+  // point from anywhere in the initial array would be eligible
+  // to end up in the final subsample.
+
+  if (num_samples == data_.size()) return;
+
+  auto erase_start = data_.begin();
+  uint32_t data_len = static_cast<uint32_t>(data_.size());
+  for (uint32_t i = 0; i < num_samples; ++i, ++erase_start) {
+    uint32_t j = i + random_idx(data_len - i);
+    std::swap(data_[i], data_[j]);
+  }
+
+  data_.erase(erase_start, data_.end());
+}
+
+template<typename T, typename A>
+template<typename FwdItem>
+void ebpps_sample<T,A>::set_partial(FwdItem&& item) {
+  if (partial_item_)
+    *partial_item_ = conditional_forward<FwdItem>(item);
+  else
+    partial_item_.emplace(conditional_forward<FwdItem>(item));
+}
+
+template<typename T, typename A>
+void ebpps_sample<T,A>::move_one_to_partial() {
+  size_t idx = random_idx(static_cast<uint32_t>(data_.size()));
+  // swap selected item to end so we can delete it easily
+  size_t last_idx = data_.size() - 1;
+  if (idx != last_idx) {
+    std::swap(data_[idx], data_[last_idx]);
+  }
+
+  set_partial(std::move(data_[last_idx]));
+
+  data_.pop_back();
+}
+
+template<typename T, typename A>
+void ebpps_sample<T,A>::swap_with_partial() {
+  if (partial_item_) {
+    size_t idx = random_idx(static_cast<uint32_t>(data_.size()));
+    std::swap(data_[idx], *partial_item_);
+  } else {
+    move_one_to_partial();
+  }
+}
+
+template<typename T, typename A>
+void ebpps_sample<T,A>::reset() {
+  c_ = 0.0;
+  partial_item_.reset();
+  data_.clear();
+}
+
+template<typename T, typename A>
+double ebpps_sample<T,A>::get_c() const {
+  return c_;
+}
+
+template<typename T, typename A>
+auto ebpps_sample<T,A>::get_full_items() const -> result_type {
+  return result_type(data_);
+}
+
+template<typename T, typename A>
+bool ebpps_sample<T,A>::has_partial_item() const {
+  return bool(partial_item_);
+}
+
+template<typename T, typename A>
+T ebpps_sample<T,A>::get_partial_item() const {
+  if (!partial_item_) throw std::runtime_error("Call to get_partial_item() when no partial item exists");
+  return *partial_item_;
+}
+
+template<typename T, typename A>
+uint32_t ebpps_sample<T,A>::random_idx(uint32_t max) {
+  std::uniform_int_distribution<uint32_t> dist(0, max - 1);
```

Comment:
I am afraid this might be expensive. is it avoidable?

Reply:
I think this should be better:
static std::uniform_int_distribution<uint32_t> dist;
return dist(random_utils::rand, std::uniform_int_distribution<uint32_t>::param_t(0, max - 1));

Reply:
Seems like it should be, yes